### PR TITLE
Feature/dfe 1027 add api for creating new release

### PR DESCRIPTION
**Pull request by richardpopplehive**

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using Xunit;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Tests.Services.DbUtils;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Tests.Services
+{
+    public class ReleaseServiceTests
+    {
+        [Fact]
+        public void GetContacts()
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs]: I think this should live in its own test class for the contact service

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleaseController.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    public class ReleasesController : ControllerBase
+    {
+        private readonly IReleaseService _releaseService;
+
+        public ReleasesController(IReleaseService releaseService)
+        {
+            _releaseService = releaseService;
+        }
+
+        // POST api/contacts
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleaseController.cs]: this comment is incorrect

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs [hunk 3]
```diff
@@ -7,6 +7,7 @@
 using AutoMapper;
 using GovUk.Education.ExploreEducationStatistics.Admin.Models.Api;
 using Microsoft.EntityFrameworkCore;
+using static GovUk.Education.ExploreEducationStatistics.Admin.Services.ReleaseService;
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs]: not so keen on the static reference here

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs [hunk 4]
```diff
@@ -39,31 +40,23 @@ public List<Publication> List()
...
                 .Include(p => p.Methodologies)
                 .ToList();
-            
-            return mapper.Map<List<PublicationViewModel>>(publications);
+
+            return PublicationToPublicationViewModelMapper.Map<List<PublicationViewModel>>(publications);
         }
         
-        private bool IsLatestRelease(Release release)
+        private static readonly IMapper PublicationToPublicationViewModelMapper = new MapperConfiguration(cfg =>
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs]: Suggest we move this out into its own mapper class

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs [hunk 5]
```diff
@@ -29,5 +35,66 @@ public List<Release> List()
         {
             return _context.Releases.ToList();
         }
+
+        // TODO Authorisation will be required when users are introduced
+        public ReleaseViewModel GetViewModel(Guid id)
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs]: Id look at asyncing this method

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs [hunk 6]
```diff
@@ -29,5 +35,66 @@ public List<Release> List()
...
+        {
+            return ReleaseToReleaseViewMapper.Map<ReleaseViewModel>(Get(id));
+        }
+
+        // TODO Authorisation will be required when users are introduced
+        public ReleaseViewModel CreateRelease(CreateReleaseViewModel createRelease)
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs]: Id look at asyncing this method

File: src/GovUk.Education.ExploreEducationStatistics.Content.Api.Tests/Controllers/ContentControllerTests.cs [hunk 7]
```diff
@@ -170,7 +169,7 @@ public void Get_Release_Returns_Ok()
             var result = controller.GetRelease("publication-a");
 
             Assert.IsAssignableFrom<Release>(result.Value);
-            Assert.Equal("Publication A", result.Value.Title);
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Content.Api.Tests/Controllers/ContentControllerTests.cs]: We should still check the title in addition to the slug

File: src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs [hunk 8]
```diff
@@ -976,13 +975,12 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                             Order = 0,
                             PublicationId = new Guid("cbbd299f-8297-44bc-92ac-558bcf51f8ad"),
                             Published = new DateTime(2018, 3, 22, 0, 0, 0, 0, DateTimeKind.Unspecified),
-                            ReleaseName = "2016 to 2017",
+                            ReleaseName = "2016",
```
Comment (jamescutts) [inline src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs]: I dont think we should be changing this, the content has been user tested, and both API and UI tests will be looking for 2016 to 2017.

Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs]: It looks like this has changed - the releaseName is basically just the year and the actual name to display to the users is the time identifier and the year. According to design anyway.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs (hunk 1 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseServiceTests.cs]: Not sure why this is here - I found a copy in the ContactServiceTests. Over zealous ctrl C ctrl V.

Removed

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleaseController.cs (hunk 2 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleaseController.cs]: Cheers - fixed

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs (hunk 3 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs]: Gone

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs (hunk 5 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs]: Done

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs (hunk 6 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseService.cs]: Done

File: src/GovUk.Education.ExploreEducationStatistics.Content.Api.Tests/Controllers/ContentControllerTests.cs (hunk 7 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api.Tests/Controllers/ContentControllerTests.cs]: Done - title its now generated from other fields

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs (hunk 4 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/PublicationService.cs]: Done

File: src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs (hunk 8 above)
Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs]: Spoken to Mark - sounds like there might be an outstanding ticket to change this to just year anyway. He is happy for this to get pushed in and to make any changes to broken tests

Comment (jamescutts) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs]: I think we should probably grab laura and confirm this, all current publications on gov.uk follow the `A to B` format e.g. https://www.gov.uk/government/statistics/pupil-absence-in-schools-in-england-2014-to-2015 so i'm still a bit cautious of us making this change.

Comment (richardpopplehive) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Api/Migrations/ApplicationDbContextModelSnapshot.cs]: changes now made to Freds requirement of e.g. 2019/20

**Pull request merged**